### PR TITLE
FbBlobStream

### PR DESCRIPTION
**Pull request by la-we**
Added FbBlobStream to be able to stream a blob from the database rather than having to fetch it completely. Especially useful when dealing with large BLOBs.

This is a working but possibly not clean solution. Have a look whether something should be cleaned up or improved.

### CODE REVIEW

**Comment by cincuranet** [comment]
Nice. Please give me some time to review it, as I'm currently busy.

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+			{
+				if (chunk > count)
+				{
+					chunk = count;
+					tmpBuffer = new byte[chunk];
+				}
+
+				Array.Copy(buffer, offset, tmpBuffer, 0, chunk);
+				await _blobHandle.PutSegmentAsync(tmpBuffer, cancellationToken)
+					.ConfigureAwait(false);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs:192]: nit: Put it on one line.

File: src/FirebirdSql.Data.FirebirdClient.Tests/FbBlobStreamTests.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+				await insert.ExecuteNonQueryAsync();
+			}
+
+			await using (var select = new FbCommand($"SELECT blob_field FROM TEST WHERE int_field = {id_value}", Connection, transaction))
+			{
+				await using var reader = await select.ExecuteReaderAsync();
+				while (await reader.ReadAsync())
+				{
+					await using var stream = reader.GetStream(0);
+					await stream.WriteAsync(insert_values);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient.Tests/FbBlobStreamTests.cs:72]: This - that you can modify the blob as part of selecting - is going to be nice expert level feature.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 3]
```diff
@@ -117,6 +121,86 @@ protected override async ValueTask OpenAsync(CancellationToken cancellationToken
...
+			if (!IsOpen)
+				Open();
+
+			var bufferLength = 20;
+			var buffer = new byte[bufferLength];
+
+			_database.Xdr.Write(IscCodes.op_info_blob);
+			_database.Xdr.Write(_blobHandle);
+			_database.Xdr.Write(0);
+			_database.Xdr.WriteBuffer(new byte[]{IscCodes.isc_info_blob_total_length}, 1);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:137]: nit: Spaces around `{` and `}`.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 4]
```diff
@@ -117,6 +121,86 @@ protected override async ValueTask OpenAsync(CancellationToken cancellationToken
...
+			if (!IsOpen)
+				await OpenAsync(cancellationToken).ConfigureAwait(false);
+
+			var bufferLength = 20;
+			var buffer = new byte[bufferLength];
+
+			await _database.Xdr.WriteAsync(IscCodes.op_info_blob, cancellationToken).ConfigureAwait(false);
+			await _database.Xdr.WriteAsync(_blobHandle, cancellationToken).ConfigureAwait(false);
+			await _database.Xdr.WriteAsync(0, cancellationToken).ConfigureAwait(false);
+			await _database.Xdr.WriteBufferAsync(new byte[]{IscCodes.isc_info_blob_total_length}, 1, cancellationToken).ConfigureAwait(false);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:177]: nit: Spaces around `{` and `}`.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 5]
```diff
@@ -216,7 +300,142 @@ protected override async ValueTask GetSegmentAsync(Stream stream, CancellationTo
 		}
 	}
 
-	protected override void PutSegment(byte[] buffer)
+	internal override byte[] GetSegment()
+	{
+		var requested = SegmentSize;
+
+#if !(NET48 || NETSTANDARD2_0)
+		var tmp = ArrayPool<byte>.Shared.Rent(requested * 2);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:308]: I would remove the `ArrayPool` from this PR. It's not used anywhere, yet. It might be bigger sweep later across the codebase.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 6]
```diff
@@ -216,7 +300,142 @@ protected override async ValueTask GetSegmentAsync(Stream stream, CancellationTo
...
+			ArrayPool<byte>.Shared.Return(tmp);
+#endif
+		}
+	}
+	internal override async ValueTask<byte[]> GetSegmentAsync(CancellationToken cancellationToken = default)
+	{
+		var requested = SegmentSize;
+
+#if !(NET48 || NETSTANDARD2_0)
+		var tmp = ArrayPool<byte>.Shared.Rent(requested * 2);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:375]: I would remove the `ArrayPool` from this PR. It's not used anywhere, yet. It might be bigger sweep later across the codebase.

File: src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs [hunk 7]
```diff
@@ -149,9 +155,55 @@ protected override ValueTask OpenAsync(CancellationToken cancellationToken = def
...
+	{
+		ClearStatusVector();
+
+		var buffer = new byte[20];
+
+		_database.FbClient.isc_blob_info(
+			_statusVector,
+			ref _blobHandle,
+			1,
+			new byte[]{IscCodes.isc_info_blob_total_length},
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:173]: nit: Spaces around `{` and `}`.

File: src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs [hunk 8]
```diff
@@ -149,9 +155,55 @@ protected override ValueTask OpenAsync(CancellationToken cancellationToken = def
...
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:195]: nit: Spaces around `{` and `}`.

**Review by cincuranet — CHANGES_REQUESTED** [review]

File: src/FirebirdSql.Data.FirebirdClient.Tests/FbBlobStreamTests.cs (hunk 2 above)
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient.Tests/FbBlobStreamTests.cs:72]: This - that you can modify the blob as part of selecting - is going to be nice expert level feature.

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs [hunk 9]
```diff
@@ -0,0 +1,229 @@
+﻿using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using FirebirdSql.Data.Common;
+
+namespace FirebirdSql.Data.FirebirdClient;
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs]: You are right in this regard, moved and renamed the class.

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+public class FbBlobStream : Stream
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs [hunk 11]
```diff
@@ -203,26 +206,32 @@ public async ValueTask WriteAsync(byte[] buffer, int index, int count, Cancellat
...
 
-	protected abstract void Close();
-	protected abstract ValueTask CloseAsync(CancellationToken cancellationToken = default);
+	internal abstract void Close();
+	internal abstract ValueTask CloseAsync(CancellationToken cancellationToken = default);
 
-	protected abstract void Cancel();
-	protected abstract ValueTask CancelAsync(CancellationToken cancellationToken = default);
+	internal abstract void Cancel();
+	internal abstract ValueTask CancelAsync(CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs [hunk 12]
```diff
@@ -29,14 +29,17 @@ internal abstract class BlobBase
...
 	protected TransactionBase _transaction;
 
 	public abstract int Handle { get; }
 	public long Id => _blobId;
 	public bool EOF => (_rblFlags & IscCodes.RBL_eof_pending) != 0;
+	public bool IsOpen => _isOpen;
 
-	protected int SegmentSize => _segmentSize;
+	internal int SegmentSize => _segmentSize;
+	internal int Position => _position;
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+#if !(NET48 || NETSTANDARD2_0)
+	public override ValueTask DisposeAsync()
+	{
+		return _blobHandle.CloseAsync();
+	}
+#endif
+
+	private static void ValidateBufferSize(byte[] buffer, int offset, int count)
+	{
+		if (buffer is null) throw new ArgumentNullException(nameof(buffer));
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+		if (buffer.Length < offset + count) throw new InvalidOperationException();
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+			}
+
+			if (_blobHandle.EOF)
+				break;
+
+			if (Available == 0)
+			{
+				_currentSegment = await _blobHandle
+					.GetSegmentAsync(cancellationToken)
+					.ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs (hunk 1 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/FirebirdClient/FbBlobStream.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs (hunk 3 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs (hunk 4 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs (hunk 5 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs (hunk 6 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs (hunk 7 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:173]: addressed

File: src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs (hunk 8 above)
Comment (lweberprb) [reply src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:195]: addressed

**Comment by lweberprb** [comment]
Hey Jiri,

i have addressed your feedback in this updated commit! Thanks a lot for having an in-depth look at the changes!

**Comment by cincuranet** [comment]
@la-we Sorry for not taking another look sooner. I was busy with work and stuff. I'll do another pass in upcoming days.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 16]
```diff
@@ -21,6 +21,10 @@
 using System.Threading.Tasks;
 using FirebirdSql.Data.Common;
 
+#if !(NET48 || NETSTANDARD2_0)
+using System.Buffers;
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:25]: Why is this needed? I don't see anything from `System.Buffers` used below.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 17]
```diff
@@ -117,7 +121,87 @@ protected override async ValueTask OpenAsync(CancellationToken cancellationToken
...
+			if (!IsOpen)
+				Open();
+
+			var bufferLength = 20;
+			var buffer = new byte[bufferLength];
+
+			_database.Xdr.Write(IscCodes.op_info_blob);
+			_database.Xdr.Write(_blobHandle);
+			_database.Xdr.Write(0);
+			_database.Xdr.WriteBuffer(new byte[]{ IscCodes.isc_info_blob_total_length }, 1);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:137]: Missing space between `[]{`.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 18]
```diff
@@ -117,7 +121,87 @@ protected override async ValueTask OpenAsync(CancellationToken cancellationToken
...
+			if (!IsOpen)
+				await OpenAsync(cancellationToken).ConfigureAwait(false);
+
+			var bufferLength = 20;
+			var buffer = new byte[bufferLength];
+
+			await _database.Xdr.WriteAsync(IscCodes.op_info_blob, cancellationToken).ConfigureAwait(false);
+			await _database.Xdr.WriteAsync(_blobHandle, cancellationToken).ConfigureAwait(false);
+			await _database.Xdr.WriteAsync(0, cancellationToken).ConfigureAwait(false);
+			await _database.Xdr.WriteBufferAsync(new byte[]{ IscCodes.isc_info_blob_total_length }, 1, cancellationToken).ConfigureAwait(false);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:177]: Missing space between `[]{`.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 19]
```diff
@@ -216,7 +300,120 @@ protected override async ValueTask GetSegmentAsync(Stream stream, CancellationTo
...
+			else if (response.ObjectHandle == 2)
+			{
+				RblAddValue(IscCodes.RBL_eof_pending);
+			}
+
+			var buffer = response.Data;
+
+			if (buffer.Length == 0)
+			{
+				// previous	segment	was	last, this has no data
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:387]: Fix spaces in comment.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs [hunk 20]
```diff
@@ -249,14 +446,14 @@ protected override async ValueTask PutSegmentAsync(byte[] buffer, CancellationTo
 		}
 	}
 
-	protected override void Seek(int position)
+	public override void Seek(int offset, int fbSeekMode)
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs:449]: `fbSeekMode` ==> `seekMode`

File: src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs (hunk 7 above)
Comment (cincuranet) [reply src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:173]: I don't see the spaces added. 🤷‍♂️

File: src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs (hunk 8 above)
Comment (cincuranet) [reply src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:195]: I don't see the spaces added. 🤷‍♂️

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs [hunk 21]
```diff
@@ -51,6 +54,7 @@ public string ReadString()
 		var buffer = Read();
 		return _charset.GetString(buffer, 0, buffer.Length);
 	}
+
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs:57]: Keep the sync and async overloads together.

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs [hunk 22]
```diff
@@ -83,6 +87,7 @@ public byte[] Read()
 			return ms.ToArray();
 		}
 	}
+
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs:90]: Keep the sync and async overloads together.

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs [hunk 23]
```diff
@@ -165,7 +172,9 @@ public void Write(byte[] buffer, int index, int count)
 			throw;
 		}
 	}
-	public async ValueTask WriteAsync(byte[] buffer, int index, int count, CancellationToken cancellationToken = default)
+
+	public async ValueTask WriteAsync(byte[] buffer, int index, int count,
+		CancellationToken cancellationToken = default)
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs:177]: No wrapping.

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs [hunk 24]
```diff
@@ -203,26 +212,32 @@ public async ValueTask WriteAsync(byte[] buffer, int index, int count, Cancellat
...
 
-	protected abstract void PutSegment(byte[] buffer);
-	protected abstract ValueTask PutSegmentAsync(byte[] buffer, CancellationToken cancellationToken = default);
+	public abstract void PutSegment(byte[] buffer);
+	public abstract ValueTask PutSegmentAsync(byte[] buffer, CancellationToken cancellationToken = default);
 
-	protected abstract void Seek(int position);
-	protected abstract ValueTask SeekAsync(int position, CancellationToken cancellationToken = default);
+	public abstract void Seek(int offset, int fbSeekMode);
+	public abstract ValueTask SeekAsync(int offset, int fbSeekMode, CancellationToken cancellationToken = default);
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobBase.cs:234]: `fbSeekMode` ==> `seekMode`

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+
+		return copied;
+	}
+
+	public override long Seek(long offset, SeekOrigin origin)
+	{
+		if (!_blobHandle.IsOpen)
+			_blobHandle.Open();
+
+		var fbSeekMode = origin switch
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs:116]: `fbSeekMode` ==> `seekMode`

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs [hunk 26]
```diff
@@ -0,0 +1,228 @@
...
+				await _blobHandle.PutSegmentAsync(tmpBuffer, cancellationToken).ConfigureAwait(false);
+
+				offset += chunk;
+				count -= chunk;
+				_position += chunk;
+			}
+		}
+		catch
+		{
+			// Cancel the blob and rethrow the exception
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs:197]: Remove the comment. It's obvious from the code.

**Review by cincuranet — CHANGES_REQUESTED** [review]
In general looks good. Just some small nits.

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs [hunk 27]
```diff
@@ -0,0 +1,228 @@
...
+			_blobHandle.Cancel();
+
+			throw;
+		}
+	}
+
+	public override async Task WriteAsync(byte[] buffer, int offset, int count,
+		CancellationToken cancellationToken)
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs:169]: Single line.

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs (hunk 4 above)
Comment (cincuranet) [reply src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs]: I don't see the spaces added. 🤷‍♂️

File: src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs (hunk 3 above)
Comment (cincuranet) [reply src/FirebirdSql.Data.FirebirdClient/Client/Managed/Version10/GdsBlob.cs]: I don't see the spaces added. 🤷‍♂️

**Comment by fdcastel** [comment]
@la-we could you rebase with the latest `master` and work on the requested changes?

This would be a nice addition to the project.

**Comment by la-we** [comment]
@cincuranet Addressed your feedback. Sorry for the long waiting period...
@fdcastel Done!

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+
+				offset += chunk;
+				count -= chunk;
+				_position += chunk;
+			}
+		}
+		catch
+		{
+			_blobHandle.Cancel();
+
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs:161]: Remove empty line.

File: src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+				offset += chunk;
+				count -= chunk;
+				_position += chunk;
+			}
+		}
+		catch
+		{
+			await _blobHandle.CancelAsync(cancellationToken).ConfigureAwait(false);
+
```
Comment (cincuranet) [inline src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs:194]: Remove empty line.

**Review by cincuranet — CHANGES_REQUESTED** [review]

**Comment by cincuranet** [comment]
@la-we Few pending comments left, mostly formatting. Otherwise LGTM and I'll merge after these are addressed.

**Comment by cincuranet** [comment]
@la-we Are you aware of the failing test?

```text
1) Error : FirebirdSql.Data.FirebirdClient.Tests.BlobStreamTests(Embedded,False,Disabled).FbBlobStreamReadTest
FirebirdSql.Data.Common.IscException : invalid BLOB handle
   at FirebirdSql.Data.Client.Native.StatusVectorHelper.ProcessStatusVector(IntPtr[] statusVector, Charset charset, Action`1 warningMessage) in /_/src/FirebirdSql.Data.FirebirdClient/Client/Native/StatusVectorHelper.cs:line 39
   at FirebirdSql.Data.Client.Native.FesDatabase.ProcessStatusVector(IntPtr[] statusVector) in /_/src/FirebirdSql.Data.FirebirdClient/Client/Native/FesDatabase.cs:line 393
   at FirebirdSql.Data.Client.Native.FesBlob.CloseAsync(CancellationToken cancellationToken) in /_/src/FirebirdSql.Data.FirebirdClient/Client/Native/FesBlob.cs:line 433
   at FirebirdSql.Data.Common.BlobStream.DisposeAsync() in /_/src/FirebirdSql.Data.FirebirdClient/Common/BlobStream.cs:line 209
   at FirebirdSql.Data.FirebirdClient.Tests.BlobStreamTests.FbBlobStreamReadTest() in /_/src/FirebirdSql.Data.FirebirdClient.Tests/BlobStreamTests.cs:line 42
   at FirebirdSql.Data.FirebirdClient.Tests.BlobStreamTests.FbBlobStreamReadTest() in /_/src/FirebirdSql.Data.FirebirdClient.Tests/BlobStreamTests.cs:line 46
   at FirebirdSql.Data.FirebirdClient.Tests.BlobStreamTests.FbBlobStreamReadTest() in /_/src/FirebirdSql.Data.FirebirdClient.Tests/BlobStreamTests.cs:line 37
   at FirebirdSql.Data.FirebirdClient.Tests.BlobStreamTests.FbBlobStreamReadTest() in /_/src/FirebirdSql.Data.FirebirdClient.Tests/BlobStreamTests.cs:line 48
```

**Comment by la-we** [comment]
Nope, did not notice to be honest. I'll have a look at the failing tests

**Comment by la-we** [comment]
@cincuranet Tests are fixed!
